### PR TITLE
Make anchors "^" and "$" optional

### PR DESCRIPTION
**Pull request by ildar-shaimordanov**
Additional options:
`-B`, `--match-beginning` - Match the beginning of the string (prepend `^`)
`-E`, `--match-end` - Match the end of the string (append `$`)
`-X`, `--match-line` - Match the whole string (as a shorthand for `-B -E`)

It's result of the discussion in the issue pemistahl/grex#30. 
Sorry, if some of my modifications look silly. It's my attempt to understand Rust from the scratch.

### CODE REVIEW

**Review by pemistahl — CHANGES_REQUESTED** [review]
Hi Ildar,

thank you for this pull request and sorry for my late response. Before I can think about merging, two points need to be addressed:

1. The anchors should still remain the default as it is now. So *disabling* the anchors should be made optional, not enabling them.
2. Because of your changes, the property tests now fail. You must make sure that they still pass when disabling the anchors.

Please tell me whether you want to work on these things. If not, I will close your PR. Thank you.

**Comment by ildar-shaimordanov** [comment]
Hi Peter,

I understand that you are main developer of the tool and your decision is final on how it would be developed further. However, setting anchors optional is less strict usage, and makes the tool more flexible.

As you can see I made 3 commits (the second one fixed some silliness of the first one, and with the third commit I began to adapt tests to new approach).

Before we decide what to do with this PR -- continue or close -- please let me know your final decision: anchors by default are optional or mandatory. It will lead to two different works: with optional anchors we need to complete tests; with mandatory anchors we need completely rework changes, revert tests and add new tests for disabled anchors.

Ildar

**Comment by pemistahl** [comment]
Hi Ildar, 

as I said, setting the anchors should remain the default. Why? Because the original concept of grex is to be as strict as possible and to provide a regex that *exactly* matches the test cases given as input and nothing else. This can only be guaranteed by setting the anchors. By applying command-line arguments such as `--words` or `--repetitions`, the tool's strictness is loosened and the set of matched test cases is extended. In this manner, optionally disabling the anchors is a further way to loosen the tool's strictness.

In any case, you have to add new property tests to make sure that the generated regexes without the anchors *always* match the test cases given as input.

Do you want to work on that?

**Comment by ildar-shaimordanov** [comment]
Hi Peter, Sorry not responding for long time. I am busy with main activity. When I have enough time I will start these changes I inform you in this thread.

**Pull request closed without merging**
